### ISSUE TITLE
[V3] Singleton Design Discussion

### ISSUE THREAD

**Issue by Tobotimus** [Cog-Creators/Red-DiscordBot#1800]
Red as we know it is an application with various singleton components at its core (literally, in the `core` package). What I'm referring to as singletons are components for which there exist only one instance of per application.

In the core package currently, this boils down to:
 - the bank
 - modlog
 - the Sentry manager
 - the help formatter
 - the cog manager
 - the data manager

There is also the RPC server manager, which is currently being refactored over at #1780, and of course the bot itself, however we're not really talking about the bot class itself here, we're talking about other singletons.

### Current design of these components
As of right now, the design of these singletons is inconsistent. Some use a singleton class, which is instantiated on bot initialisation and attached to the bot as an attribute. This includes sentry, the help formatter, the cog manager, and RPC.

The rest simply use their module as a container for data and as an interface to functions, which can be imported at the top of a module. Python allows us to do this as an alternative to the more "Java-esque" approach of singleton classes. But is it good practice? Well, that's what this issue is about.

#### The singleton class approach
Personally, I see this approach as simpler, easier to maintain, and more extendable. On the other hand, it can be seen as polluting the bot object's attributes. It also means we need access to the bot in order to have access to these components, but thankfully the bot object is very easy to get a hold of.

For the help formatter, this approach is actually forced upon us by discord.py and is necessary to allow anyone to customise the help formatter, hence why I mentioned this approach being more extendable.

#### The singleton module approach
This approach is fairly easy to document and use, as users only need to do an import and they can access all of these singleton components. However, this does lead to more imports than may be necessary, and sometimes I fear one day we're gonna fall upon some horrible circular import problems  within the core package.

### What's better?
Personally I prefer the former approach, but this issue is for others to chip in and add to what I have stated above so we can have some sort of standard moving forward.

#### So when we decide what's better, we should rewrite everything to conform to that?
In short, no, not simply for the sake of conforming to a new standard. There may be a reason down the line for us to switch existing singletons over, but right now it's a risk which is certainly not worth taking - all of the singletons currently function just fine, rewriting them will likely just cause more issues, and we also just don't need the extra work. However, hopefully the outcome of this discussion will serve as a good reference for any future developments.

**Comment by tekulvw** [comment]
I'd like to propose a third, hybrid option.

The module approach was used for the bank specifically because it did not need to be attached to the bot and made it much easier to write code for Economy and third party cogs if the reference to the bank functions was made available via import. I am a huge fan of making objects available via import because it makes them far easier (and in my opinion, more pythonic) to access than having to maintain a reference to the bot object everywhere you might need any of these other framework components.

However, I have realized that a pure module based approach does lead to module pollution and that's normally a bad thing for numerous reasons. As a result, I'm in agreement that everything should be encapsulated in classes but I would still like access to objects from the framework via import.

As a result, I would like to propose a `redbot.registry` module where we can keep references to all of the global singleton objects as listed above. For example, to access the bank, we would now be required to do:

```py
from redbot import registry

await registry.bank.METHOD()
```

**Comment by mikeshardmind** [comment]
I'd like to point out one other possible option which is already in use (sort of) in the bot code. see `redbot.core.utils.tunnel`

We don't need a global reference to these things, we can use metaclasses and weakref to return the existing object when instanced if one already does, ensuring only one of them ever exists.

**Comment by Tobotimus** [comment]
I'm not sold on having a module containing global references to singletons. I don't particularly see a downside in requiring a reference to the bot object in order to access them.

Singletons act as interfaces to the application's shared state. This state is largely held in the bot object. I think that holding a reference to the bot object is a simple, intuitive (and I might say "pythonic") way of mutating that shared state.

**Comment by tekulvw** [comment]
My argument against having the bot class hold references to everything is that I don't want it turning into a god class. We are developing a framework around the bot with (somewhat) independent features that don't require the bot object to operate and I feel, in a design sense, they should be available without access to the bot object.

Yes, every piece of code we write acts as an interface to the *application's* shared state. The bot obect is the interface to *discord's* shared state. I would argue that our application is no longer just *discord* shared state. As an example, while the bank requires discord state information to be passed into it, it doesn't require knowledge or manipulation of that state internally. It is operating as an independent "service" or feature to *discord's* shared state.

**Comment by Tobotimus** [comment]
Right, I can see where you're coming from now, and I can agree with your sentiments.

Not sure about the breaking changes which would be introduced by a `registry` module, it's either live with a bit of import/module pollution or fix it all now and put it into the registry.

I guess that by this standard, the help formatter ought to be the only singleton attached to the bot. The rest can more or less function independently, and if they require the bot, it'd only be a one-way aggregation. This could also create breaking changes for some singletons such as modlog, which requires the bot object to be passed in for most of its methods.

**Comment by mikeshardmind** [comment]
If we're not attaching to the bot, this is a case where we could still use a central method of ensuring certain methods remain singletons without requiring global access (directly) to the objects, This single metaclass definition, and assigning what needs to be a singleton this as it's metaclass would effectively wrap all those classes in the equivalent of a SingletonFactory in Java (going back to Java design comparisons from earlier)

A singleton metaclass for the entirety of the bot could handle this
```py
import weakref

_singletons = weakref.WeakValueDictionary({})


class Singleton(type):

    def __call__(cls, *args, **kwargs):
        print(cls.__name__)
        if cls.__name__ in _singletons:
            return _singletons[cls.__name__]
        temp = super(Singleton, cls).__call__(*args, **kwargs)
        _singletons[cls.__name__] = temp
        return temp
```

**Comment by tekulvw** [comment]
I don't believe that code to enforce single object instantiation is necessary as there's absolutely no reason why we would ever re-instantiate the singletons. They are created in a non-public part of the API as part of the bootstrapping process within the bot.

Any singleton that doesn't require a reference to the bot object would simply be instantiated at runtime and the rest would be instantiated in an `initialize` function called once right after the bot object is created.

**Comment by mikeshardmind** [comment]
That's a fair assessment based on the current singletons, but having it could also simplify safe access to them without needing to attach them to the bot. It's a thought and nothing more on this, and it probably is overengineering it at the current state.

**Comment by Tobotimus** [comment]
I'm sold on @tekulvw's proposal as to how we should be moving forward, and for that I'm gonna close this issue (also closing due to inactivity). If we have a change of heart it can be reopened.
